### PR TITLE
Publish steps when publishing workflow

### LINKED ISSUE #100: Publish steps when publishing workflow
If creating a workflow from scratch like so:
```python
    workflow = FairWorkflow(description= 'This is a test workflow.', label='Test')

    step1 = FairStep(uri='http://www.example.org/step1')
    step2 = FairStep(uri='http://www.example.org/step2')
    step3 = FairStep(uri='http://www.example.org/step3')

    workflow.first_step = step1
    workflow.add(step2, follows=step1)
    workflow.add(step3, follows=step2)
```

And then publishing it:
```python
workflow.publish_as_nanopub(use_test_server=True)
```
We end up with [this published workflow](http://test-server.nanopubs.lod.labs.vu.nl/RAQ3H8K2RM_45AN2qgTX_DGB842395SFrkZi0ggeYpzDA).
The steps are not published.

- Publish modified steps before publishing workflow
- Use URIs of published steps in workflow before publishing workflow

### PR DESCRIPTION
**Pull request by svenvanderburg**
Fixes #100

### CODE REVIEW

File: tests/test_fairworkflow.py [hunk 1]
```diff
@@ -14,17 +14,39 @@
 class TestFairWorkflow:
```
Comment (svenvanderburg) [inline tests/test_fairworkflow.py:14]: Sorry for all the unrelated changes regarding fixtures here... I had to fix this to make the tests not mutate the objects (making the test success dependent on which tests are run before it).

Comment (svenvanderburg) [reply tests/test_fairworkflow.py:14]: Only the last two tests (scroll down) were actually changed.

File: fairworkflows/fairworkflow.py [hunk 2]
```diff
@@ -414,6 +414,28 @@ def draw(self, filepath):
...
+            a dictionary with publication info, including 'nanopub_uri', and 'concept_uri' of the
+                published workflow
+        """
+        for step in self:
+            if step.is_modified or not step._is_published:
+                self._is_modified = True  # If one of the steps is modified the workflow is too.
+                old_uri = step.uri
+                step.publish_as_nanopub(use_test_server=use_test_server)
+                published_step_uri = step.uri
+                replace_in_rdf(self.rdf, oldvalue=rdflib.URIRef(old_uri),
```
Comment (raar1) [inline fairworkflows/fairworkflow.py:433]: Nice - this was always a tricky thing for this work. Having to carry around URIs that will change when published (to unpredictable URIs).

File: fairworkflows/fairworkflow.py [hunk 3]
```diff
@@ -414,6 +414,28 @@ def draw(self, filepath):
                 'Cannot produce visualization of RDF, you need to install '
                 'graphviz dependency https://graphviz.org/')
 
+    def publish_as_nanopub(self, use_test_server=False):
```
Comment (raar1) [inline fairworkflows/fairworkflow.py:417]: I'm thinking that we should probably have a ```publish_all_steps``` kwarg here too that is ```True``` by default, as it has been raised in previous discussions that some users may not want to publish all steps (FAIR doesn't mean open, etc etc). In such a case a user may wish to control which they do and don't publish.

File: tests/test_fairworkflow.py (hunk 1 above)
Comment (raar1) [reply tests/test_fairworkflow.py:14]: Clearly this is totally unacceptable behaviour, Sven, but I will let it pass this one time.

File: tests/test_fairworkflow.py [hunk 4]
```diff
@@ -287,25 +305,64 @@ def test_draw_with_graphviz_module_without_dependency(self, tmp_path):
...
+                               'www.example.org/published_workflow#workflow']
+        mock_publish.side_effect = [
+            {'concept_uri': test_published_uris[0]},  # first call
+            {'concept_uri': test_published_uris[1]},
+            {'concept_uri': test_published_uris[2]},
+            {'concept_uri': test_published_uris[3]}   # Last call
+        ]
+        for step in test_workflow:
+            assert step.is_modified
+        test_workflow.publish_as_nanopub()
```
Comment (raar1) [inline tests/test_fairworkflow.py:344]: Do we need to check the returned publication_info too?

**Review by raar1 — APPROVED** [review]
All looks good to me. I've left a couple of comments.

File: tests/test_fairworkflow.py (hunk 1 above)
Comment (svenvanderburg) [reply tests/test_fairworkflow.py:14]: Lol

File: fairworkflows/fairworkflow.py (hunk 3 above)
Comment (svenvanderburg) [reply fairworkflows/fairworkflow.py:417]: I agree! How to solve that though? We need some way of referencing the step without it being published. A blank node saying 'unpublished step', or maybe there is something in the pplan ontology that we can refer to?

File: tests/test_fairworkflow.py (hunk 4 above)
Comment (svenvanderburg) [reply tests/test_fairworkflow.py:344]: No, because the publish method is mocked and publication_info is just what we tell it to return here:
https://github.com/fair-workflows/fairworkflows/blob/6e58c8a4911b90b4d591aa68530f07499757611b/tests/test_fairworkflow.py#L336

Comment (svenvanderburg) [reply tests/test_fairworkflow.py:344]: Do we need to make this more clear?

File: fairworkflows/fairworkflow.py (hunk 3 above)
Comment (raar1) [reply fairworkflows/fairworkflow.py:417]: Could maybe raise this with Tobias. A blank node with a 'unique' name (so that it gets tranformed by ```nanopub``` into PUBLISHED_NANOPUB_URI#blanknodename as usual) might be enough? Alternatively we can make this an issue of its own and come back to it.

File: tests/test_fairworkflow.py (hunk 4 above)
Comment (raar1) [reply tests/test_fairworkflow.py:344]: Ah no, sorry, I hadn't noticed it was mocked. I suppose my comment was triggered by seeing that the other new test (```test_publish_as_nanopub_no_modifications()```) does check the returned publication_info whereas this test doesn't. Both tests have a mocked client so it just looked asymmetric. Not really a big issue or anything.

Comment (svenvanderburg) [reply tests/test_fairworkflow.py]: Well, I added the assertion anyway :) It is kind of trivial, but a bug could in theory be introduced there. And at least the confusion that you had wouldn't happen again!

File: fairworkflows/fairworkflow.py (hunk 3 above)
Comment (svenvanderburg) [reply fairworkflows/fairworkflow.py:417]: I've created #106 for this.
